TLS server improvements

ysbaddaden opened the pull request:
This allows to configure an OpenSSL context suitable for HTTP2. Most specifically we now can:

- Select the SSL/TLS method (sslv2, sslv3, tls1, tls1.1, tls1.2 (default));
- Configure SSL options & modes;
- Configure ciphers to allow/disable;
- Generate an ECDH key to enable EECDH ciphers (only `NID_X9_62_prime256v1` is supported, other curves may work too);
- Negotiate the protocol through ALPN (ie. `h2` for HTTP2/TLS negotiation with browsers) which requires OpenSSL 1.0.2+ (ie. Ubuntu Xenial LTS).

It also features a default SSL context suitable for HTTP/1 that should be strong enough (AFAIK). It only misses certificates to be used:

```crystal
ssl_context = HTTP::Server.default_ssl_context
ssl_context.certificate_chain = "/path/to/server.crt"
ssl_context.private_key = "/path/to/server.key"
```

Then for HTTP2:

```crystal
ssl_context.alpn_protocol = "h2"
```

Please note that I enabled RC4 in the list of CIPHERS for HTTP. This is only required to support Internet Explorer 8. We may want to disable RC4 by default, and have a way to enable it on demand?

```crystal
HTTP::Server.default_ssl_context(rc4: true)
```

ysbaddaden commented on src/openssl/lib_ssl.cr:
```diff
@@ -38,9 +45,53 @@ lib LibSSL
+  SSL_OP_NO_SESSION_RESUMPTION_ON_RENEGOTIATION = 0x00010000_u32
+  SSL_OP_SINGLE_ECDH_USE = 0x00080000_u32
+  SSL_OP_NO_TICKET = 0x00004000_u32
+  SSL_OP_CIPHER_SERVER_PREFERENCE = 0x00400000_u32
```
Maybe the options should be an enum? `SSL::Context#options` would then return an enum and be nicer to deal with.

ysbaddaden commented on src/openssl/ssl/context.cr:
```diff
@@ -19,6 +19,63 @@ class OpenSSL::SSL::Context
+  end
+
+  def clear_options(options = 0)
+    LibSSL.ssl_ctx_ctrl(@handle, LibSSL::SSL_CTRL_CLEAR_OPTIONS, options, nil)
```
I'm not fond of the naming, and this doesn't fit quite well with Crystal. I'm open to suggestions.

What happens is that `SSL_CTRL_OPTIONS` will add options to the context, whereas `SSL_CTRL_CLEAR_OPTIONS` will first clear the existing options before setting the new ones, hence fully replacing the options.

jhass replied:
How about having `options=` fully replace and `merge_options`, `update_options` or even just `add_options`?

ysbaddaden replied:
Maybe just:

```crystal
def options=(options, replace = false)
end
```

jhass replied:
No, additional options in setters don't work IMO

ysbaddaden replied:
Errata: I misunderstood how it worked: `SSL_CTRL_CLEAR_OPTIONS` doesn't reset options, but removes the passed options. To replace options, one must (using macros for brevity):

```crystal
SSL_clear_options(ctx, SSL_get_options(ctx))
SSL_set_options(ctx, SSL_OP_NO_SSLv2)
```

Maybe we should just do that in `options=` instead of dealing with OpenSSL weirdness. Setting options will always replace them? and have a `add_options()` method?

technorama replied:
+1 for the above except `add_options` is unnecessary.  Treat it like any other enum.  `ctx.options |= LibSSL::SSL_OP_NO_SSLv2`.

jhass commented:
> Please note that I enabled RC4 in the list of CIPHERS for HTTP. This is only required to support Internet Explorer 8. We may want to disable RC4 by default, and have a way to enable it on demand?

Yes, disabling it is the saner default, especially since it's more likely that this is used for server to server connections in a reverse proxy setup.

How much of this do we want to duplicate with the client? Maybe worth doing both in one shot?

What about taking the context as a keyword argument in the `HTTP::Server` constructors?

ysbaddaden commented:
I don't know much about clients (yet). Maybe the default context can be shared? It will require some specific work, thought, specifically for validating the domain. See #2073 and https://github.com/crystal-lang/crystal/issues/2073#issuecomment-217586134. I'm not sure when I'll have time to tackle these, yet I already had the server improvements ready, so here they come.

- Passing context to HTTP::Server: definitely!
- RC4: I'll disable it.
- Travis: I forgot to format `lib_ssl.cr` ...

asterite commented:
Awesome, thanks!

I personally know nothing about OpenSSL so I don't feel capable of deciding whether this is good or not. That said, we need someone who can take care of OpenSSL related stuff. We already have others PR and issues related to this:

* https://github.com/crystal-lang/crystal/issues/1710
* https://github.com/crystal-lang/crystal/pull/2180
* https://github.com/crystal-lang/crystal/pull/2343

I remember @waj also told me he doesn't know much about it. We could of course take some time to learn it, but we can't just apply this to every piece of the standard library, it would take a lot of time! :-)

@ysbaddaden Since you are sending this PR, would you like to take this role? Or maybe you and @jhass . Just verify that PR related to OpenSSL are fine according to your experience. I will take a look too, but don't wait for my confirmation to merge. Since there's still more stuff to do until we reach 1.0 it's better to move forward, even slowly. Before releasing 1.0 we can review all OpenSSL related functionality and make sure it works well and has a good API.

Otherwise, if someone else wants to take this role, let us know.

jhass commented:
Well, @datanoise would be another obvious candidate ;) Maybe you have some time to review OpenSSL related PRs? You don't have to watch all PRs, we could call you in as necessary.

ysbaddaden commented:
I don't know much about OpenSSL in particular, I follow people who do know about cryptography, and I document myself as I go. I'm trying to make the integration "correct" if not "good", provide good defaults when possible, and have a simple API.

Here I concentrate on servers, but I'll try to prove the clients too, and review/rework PR.

ysbaddaden commented:
Here comes something better. I read the OpenSSL manpages for all the methods I use and fixed the misunderstood behaviors. For example we must use `SSLv23_method` for the TLS protocol to use to be negotiated then disable SSLv2 and SSLv3 throught options (yes, not obvious). I wrote some minimal documentation, specs, and added error checking.

I finally settled for `add_options`, and `remove_options` methods (same for the modes). Having a ValueObject would look better (eg: `ctx.options.add()` and `ctx.options.add()`) but I'm not sure this is really worth it.

I modified the default HTTP server SSL context to follow the [Mozilla intermediate recommendation](https://wiki.mozilla.org/Security/Server_Side_TLS). This disables all broken ciphers (eg: RC4) as well as SSLv2 and SSv3 and should be compatible with many browsers. It's still missing the DH params for forward secrecy. I'm not sure how to generate/enable these.

jhass commented:
LGTM, nice work!

jhass commented:
Looks like there's some portability issues :)

ysbaddaden commented:
- Linux32: there seems to be a specific `SSL_OP_ALL` value for OpenSSL on 32 bits. I'll check that.
- OSX: I guess I'll have to find a list of ciphers where at least one cipher gets activated, otherwise we get that error. I suppose it uses the system's libssl instead of openssl installed by homebrew that requires link flags.

jhass commented:
Looks like CI passed, ready then?

ysbaddaden commented:
Yes, let's merge!
